saleor/graphql/invoice/mutations.py

```diff
@@ -0,0 +1,273 @@
+from ...invoice.emails import send_invoice
+from ...invoice.error_codes import InvoiceErrorCode
+from ...order import OrderStatus
+from ..core.mutations import ModelDeleteMutation, ModelMutation
+from ..core.types.common import InvoiceError
+from ..invoice.types import Invoice
+from ..order.types import Order
+
+
+class RequestInvoice(ModelMutation):
+    class Meta:
+        description = "Request an invoice for the order."
+        model = models.Invoice
+        permissions = (OrderPermissions.MANAGE_ORDERS,)
+        error_type_class = InvoiceError
+        error_type_field = "invoice_errors"
+
+    class Arguments:
+        order_id = graphene.ID(
+            required=True, description="ID of the order related to invoice."
+        )
+        number = graphene.String(
+            required=False, description="Invoice number (optional)."
+        )
+
+    @classmethod
+    def clean_instance(cls, info, instance):
+        if instance.status == OrderStatus.DRAFT:
+            raise ValidationError(
+                {
+                    "orderId": ValidationError(
+                        "Cannot request an invoice for draft order.",
+                        code=InvoiceErrorCode.INVALID_STATUS,
+                    )
+                }
+            )
+
+        if not instance.billing_address:
+            raise ValidationError(
+                {
+                    "orderId": ValidationError(
+                        "Cannot request an invoice for order without billing address.",
+                        code=InvoiceErrorCode.NOT_READY,
+                    )
+                }
+            )
+
+    @classmethod
+    def perform_mutation(cls, _root, info, **data):
+        order = cls.get_node_or_error(
+            info, data["order_id"], only_type=Order, field="orderId"
+        )
+        cls.clean_instance(info, order)
+
+        invoice = models.Invoice.objects.create(
+            order=order,
+            number=data.get("number"),
+            pending_target=PendingTarget.COMPLETE,
+        )
+
+        info.context.plugins.invoice_request(
+            order=order, invoice=invoice, number=data.get("number")
+        )
+        events.invoice_requested_event(
+            user=info.context.user, order=order, number=data.get("number")
+        )
+        invoice.refresh_from_db()
+        return RequestInvoice(invoice=invoice)
+
+
+class CreateInvoiceInput(graphene.InputObjectType):
+    number = graphene.String(required=True, description="Invoice number")
+    url = graphene.String(required=True, description="URL of an invoice to download.")
+
+
+class CreateInvoice(ModelMutation):
+    class Arguments:
+        order_id = graphene.ID(
+            required=True, description="ID of the order related to invoice."
+        )
+        input = CreateInvoiceInput(
+            required=True, description="Fields required when creating an invoice."
+        )
+
+    class Meta:
+        description = "Creates an invoice."
```

Comment:
As an API user, I see here two similar mutations - `requestInvoice` and `createInvoice`? Which one and when should I use? These descriptions don't tell that. Can we explain it shortly in both mutations?

Reply:
The same applies to `deleteInvoice` and `requestDeleteInvoice` mutations.

Reply:
Will do so, furthermore - after everything is merged to master I will spend a while and update the docs. We have a synchronous flow, async flow, plugin to handle invoices. A guide how to make use of all of this or customize it to one needs would be super handy.